src/engine/strat_engine/backstore/metadata.rs

```diff
@@ -36,7 +36,52 @@ pub struct BDA {
     regions: mda::MDARegions,
 }
 
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+enum MetadataLocation {
+    Both,
+    First,
+    Second,
+}
+
 impl BDA {
+    fn read<F>(f: &mut F) -> Result<[u8; _BDA_STATIC_HDR_SIZE]>
+    where
+        F: Read + Seek,
+    {
+        #![allow(unused_io_amount)]
+        f.seek(SeekFrom::Start(0))?;
+        let mut buf = [0u8; _BDA_STATIC_HDR_SIZE];
+
+        // FIXME: See https://github.com/stratis-storage/stratisd/pull/476
+        f.read(&mut buf)?;
+        Ok(buf)
+    }
+
+    fn write<F>(f: &mut F, bda_buf: &[u8], which: MetadataLocation) -> Result<()>
+    where
+        F: Seek + SyncAll,
+    {
+        let zeroed = [0u8; _BDA_STATIC_HDR_SIZE];
```

Comment:
This snippet of code you have placed here doesn't make sense to me.  We must have a `sync_all` between copies and the `f.write_all(if which == Both || which == First) { bda_buf } else { &zeroed...})?;` appears to zero out the metadata block.

I'm not sure I understand the value in just writing the zero parts, I think doing that increases risk of not being able to read up either copy.  If we get in the situation of having one good and one bad, we should not re-write the good one at all, even just the zero parts.

In fact I'm going to update this PR for the `BDA::read` because we are doing it incorrectly.  We should actually be issuing 2 reads of the specific BDA blocks with a seek in between instead of doing a single read.  The reason for this is that if we lose power in the middle of updating one of the copies a disk sector could be left in an error state, likely a data sync error (03/1600 which needs to be re-written to correct).  Thus a single read of the first 8K could fail with an EIO when in reality a single block read of each of the specific BDA sectors may succeed for each.  Every sector in the transfer must be good for the entire transfer to succeed, thus it's in our best interest to only read the **sectors** that really matter and read them individually.

Reply:
I've tried to achieve the same behavior as what we had before, in that we write the first copy, sync, then write the second copy with the change allowing the update of each part individually.  The individual writes done before the sync are all stream and will be coalesced into a single underlying device write so ultimately I don't think we need to change anything else here.  @agrover can speak to why we zero'd or should still zero the other blocks in the header, but in general the important part is getting one copy down intact.